pkg/cgroups/cgroupstats.go

```diff
@@ -0,0 +1,500 @@
+	)
+
+	// File looks like this:
+	//
+	// total=44611 N0=32631 N1=7501 N2=1982 N3=2497
+	// file=44428 N0=32614 N1=7335 N2=1982 N3=2497
+	// anon=183 N0=17 N1=166 N2=0 N3=0
+	// unevictable=0 N0=0 N1=0 N2=0 N3=0
+	// hierarchical_total=768133 N0=509113 N1=138887 N2=20464 N3=99669
+	// hierarchical_file=722017 N0=496516 N1=119997 N2=20181 N3=85323
+	// hierarchical_anon=46096 N0=12597 N1=18890 N2=283 N3=14326
+	// hierarchical_unevictable=20 N0=0 N1=0 N2=0 N3=20
+
+	entry := path.Join(cgroupPath, cgroupEntry)
+	lines, err := readCgroupFileLines(entry)
+	if err != nil {
+		return NumaStat{}, err
+	}
+
+	result := NumaStat{}
+	for _, line := range lines {
+		split := strings.Split(line, " ")
+		if len(line) < 2 {
+			return NumaStat{}, fmt.Errorf("error parsing file %s", entry)
+		}
+
+		keytotal := strings.Split(split[0], "=")
+		if len(keytotal) != 2 {
+			return NumaStat{}, fmt.Errorf("error parsing file %s", entry)
+		}
+		key, tot := keytotal[0], keytotal[1]
+
+		total, err := strconv.ParseInt(tot, 10, 64)
+		if err != nil {
+			return NumaStat{}, fmt.Errorf("error parsing file %s: %v", entry, err)
+		}
+
+		nodes := make(map[string]int64)
+		for _, nodeEntry := range split[1:] {
+			nodeamount := strings.Split(nodeEntry, "=")
+			if len(nodeamount) != 2 {
+				return NumaStat{}, fmt.Errorf("error parsing file %s", entry)
+			}
+			node, amount := nodeamount[0], nodeamount[1]
+			number, err := strconv.ParseInt(amount, 10, 64)
+			if err != nil {
+				return NumaStat{}, fmt.Errorf("error parsing file %s: %v", entry, err)
+			}
+			nodes[node] = number
+		}
+
+		switch key {
+		case "total":
+			result.Total.Total = total
+			result.Total.Nodes = nodes
+		case "file":
+			result.File.Total = total
+			result.File.Nodes = nodes
+		case "anon":
+			result.Anon.Total = total
+			result.Anon.Nodes = nodes
+		case "unevictable":
+			result.Unevictable.Total = total
+			result.Unevictable.Nodes = nodes
+		case "hierarchical_total":
+			result.HierarchicalTotal.Total = total
+			result.HierarchicalTotal.Nodes = nodes
+		case "hierarchical_file":
+			result.HierarchicalFile.Total = total
+			result.HierarchicalFile.Nodes = nodes
+		case "hierarchical_anon":
+			result.HierarchicalAnon.Total = total
+			result.HierarchicalAnon.Nodes = nodes
+		case "hierarchical_unevictable":
+			result.HierarchicalUnevictable.Total = total
+			result.HierarchicalUnevictable.Nodes = nodes
+		default:
+			return NumaStat{}, fmt.Errorf("error parsing file, unknown key %s", key)
+		}
+	}
+
+	return result, nil
+}
+
+// GetGlobalNumaStats returns the global (non-cgroup) NUMA statistics per node.
+func GetGlobalNumaStats() (map[int]GlobalNumaStats, error) {
+	const (
+		prefix = "/sys/devices/system/node/node"
+	)
+
+	// Files look like this:
+	//
+	// numa_hit 1851614569
+	// numa_miss 0
+	// numa_foreign 0
+	// interleave_hit 49101
+	// local_node 1851614569
+	// other_node 0
+
+	result := make(map[int]GlobalNumaStats)
+
+	nodeDirs, err := filepath.Glob(prefix + "*")
```

Comment:
My previous comment to this was a bit vague. What I tried to say (in addition to removing the regexp) is that you could get rid of the glob too by doing ReadDir in `/sys/bus/node/devices/`. This dir does not contain any additional files, just symlinks to the nodes)

Reply:
No, I noticed the comment earlier but I ignored it deliberately. It's true that it probably would be faster to do it the way you suggest (kernel do an extra symlink dereference vs. golang implementation of glob). Our other parts of the code does the same thing as here, so I left it untouched for consistency. We can change that later for all of our code.